### ISSUE TITLE
Roadmap of SoS

### ISSUE THREAD

**Issue by BoPeng** [vatlab/SoS#503]
With the task integration of Jupyter notebook, I have cleared all big features from my TODO list. My priority will now turn to stabilizing sos to reach beta 2.

Roles:

* Bo: Code optimization and bug fixes. Documenting all features.
* Gao: Test and documentation, and most importantly, manuscript
* James: Test remote execution features on various server environments.
* Jun: UI touch up and bug fix as needed

Deadlines:

* End of May: beta 2, ready of remote execution feature for a presentation on SoS.
* End of May: draft manuscript.
* End of June: submission of manuscript.
* Aug: JupyterCon presentation on SoS

**Comment by gaow** [comment]
Great -- I'll continue with the draft now and think of things we might want to add to SoS along the way! I should too get to the test of cluster jobs and Jupyter integration for my project this week.

**Comment by BoPeng** [comment]
With the new toc, the only feature I wanted is the "active markdown" (#390) cells but not sure if it can be implemented without much effort.

**Comment by jirikuncar** [comment]
Hi, I have been playing around with the [idea](https://gist.github.com/jirikuncar/e4958e2c386e070d53b23985e28a4a49) of using remote kernels running on supercomputer. Do you have any proposal how to integrate the remote ikernel in better way then it is done in https://github.com/jirikuncar/sos-remote-ikernel/?

**Comment by gaow** [comment]
Hi @jirikuncar are you asking a more generic question how to run a Jupyter notebook remotely? I use a script along the line of [this post](http://ipyrad.readthedocs.io/HPC_Tunnel.html) hope it helps. If there are further SoS specific issues it would be great if you can open a new ticket for discussions.

**Comment by BoPeng** [comment]
Is [sos-remote-ikernel](https://github.com/jirikuncar/sos-remote-ikernel/) private? I cannot access it from here (404 error). In generally, Jupyter cannot started remotely and be accessed remotely so I am not quite sure what problem this package solves.

**Comment by jirikuncar** [comment]
Hups, sorry @BoPeng, I have made it public.

**Comment by jirikuncar** [comment]
I use a bit different approach then running the whole Jupyter notebook using HPC_Tunnel. Instead I run the SoS kernel locally, but I have remote Python and R kernels running running on HPC.

**Comment by BoPeng** [comment]
I had a quick look at sos-remote-ikernel. It seems to provide wrappers to customize background colors of R and Python kernels but I cannot find anything related to HPC_Tunnel. Am I missing anything?

**Comment by jirikuncar** [comment]
@BoPeng I am taking a different approach to HPC_Tunnel. There is a nice package `remote_ikernel` which helps with creating and running remote Jupyter kernels (e.g. IPython kernel, IR). But I didn't find a way to enable easy transfer of variables (`%get x`, `%put x`) between SoS kernel and remote ones. Hence, I have created wrappers in `sos-remote-ikernel`.

**Comment by BoPeng** [comment]
I see, this model does not work for me because working nodes in our HPC do not have access to outside world (the desktop). Anyway, do `%get x` and `%put x` work once you wrap remote `sos_R` as `sos_RSlurm`? I suppose that you can start a `remote sos kernel` ticket for what you would like to achieve. Frankly I have thought of multiple kernels for one language but could not find a strong use case for it.

**Comment by jirikuncar** [comment]
@BoPeng it is about time & `$` ;) You don't want to run the main (SoS) kernel on supercomputer if you might need only one task to be executed there. `%get x` and `%put x` work well on wrapped kernels (as long as you have compatible Python/R versions running there).

Multiple kernels are useful also when you have some "old well tested" libraries which are running on Python 2.7 only, but you want to use Python 3.x for everything else.

**Comment by BoPeng** [comment]
I use `task: queue='supercomputer'` to submit large jobs to the super computer, which seems to be easier than the remote ikernel method (which needs a lot of configuration and several commands to start).

For "old well tested" libraries, sos currently supports both "Python2" and "Python3" kernels so at least this is not a concern. The only scenario I can think of for multiple kernels of the same type would be separating similar analysis on different datasets to avoid mixup of variables and data.

**Comment by jirikuncar** [comment]
But how do you work `task: queue='supercomputer'` interactively from Jupyter notebook? Maybe we are both trying to cover different use-case.

**Comment by BoPeng** [comment]
`task: queue='supercomputer'` will create one or more tasks, transfer them to `supercomputer` and start them directly or through a job queue. On the notebook, you will see something like [this](http://vatlab.github.io/SOS/doc/documentation/Notebook_Interface.html#%taskinfo-1419). The cell is not blocked so you can perform other analysis in the same notebook.

**Comment by BoPeng** [comment]
I guess this summarize the two approaches.

| | remote ikernel | sos tasks|
| --- | ---- | -----|
| computing engine | always on (living kernel) | use only when needed |
| support for task queue| no | yes |
| multiple jobs | no | yes |
| different kernels | yes, remote kernel can be any kernel or sos | no (sos wrapper of scripts in different languages)|
|multiple hosts | need to start kernel on each host | unlimited |
| local interaction | kernel blocked for large task? | tasks do not block jupyter |
| results | interactive evaluation, results as variables | results can only be returned by files |
|file synchronization| files should be on remote host | automatic synchronization of files |

**Comment by jirikuncar** [comment]
@BoPeng perfect summary! I would say that there is a place for both approaches. I see the benefits of SoS tasks in many situations and we will definitely use them in parallel to remote_ikernel solution.

**Comment by BoPeng** [comment]
I have gone through the tickets and resolved as many as I can. With the introduction of `trunk_size`, `trunk_workers` ( #602 ), and `shared` task option ( #615 ), `remote` target ( #475 ), command `resume` (#537, #611), and the sos template ( #619 ), I can conclude with some confidence that sos is now feature complete so that we can focus on the manuscript and my talk (in two weeks) as planned.

I will try to pop a champagne for myself tonight. :-)

**Comment by BoPeng** [comment]
I have finally delivered the talk and released SoS 0.9.8 today. The presentation has been uploaded to the usual place. Any update on the manuscript? @gaow ?

**Comment by gaow** [comment]
This is great! I'm going to add to the manuscript these recent changes and have an updated draft version by Monday June 12.
